src/kemal-csrf.cr

```diff
@@ -10,35 +10,34 @@ require "kemal-session"
 # Without CSRF protection, your app is vulnerable to replay attacks
 # where an attacker can re-submit a form.
 #
-class CSRF < HTTP::Handler
-  HEADER          = "X_CSRF_TOKEN"
-  ALLOWED_METHODS = %w(GET HEAD OPTIONS TRACE)
-  PARAMETER_NAME  = "authenticity_token"
+class CSRF < Kemal::Handler
+  def initialize(@header = "X_CSRF_TOKEN", @allowed_methods = %w(GET HEAD OPTIONS TRACE), @parameter_name = "authenticity_token", @error = "Forbidden")
+  end
 
   def call(context)
     unless context.session.string?("csrf")
       context.session.string("csrf", SecureRandom.hex(16))
     end
 
-    return call_next(context) if ALLOWED_METHODS.includes?(context.request.method)
+    return call_next(context) if @allowed_methods.includes?(context.request.method)
 
     req = context.request
-    submitted = if req.headers[HEADER]?
-                  req.headers[HEADER]
-                elsif context.params.body[PARAMETER_NAME]?
-                  context.params.body[PARAMETER_NAME]
+    submitted = if req.headers[@header]?
+                  req.headers[@header]
+                elsif context.params.body[@parameter_name]?
+                  context.params.body[@parameter_name]
                 else
                   "nothing"
                 end
     current_token = context.session.string("csrf")
 
     if current_token == submitted
       # reset the token so it can't be used again
-      context.session.string("csrf", SecureRandom.hex(16))
+      # context.session.string("csrf", SecureRandom.hex(16))
```

Comment:
https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet

This article suggests on csrf per session. It also helps with the back button and using csrf tokens with client side javascript.


> In general, developers need only generate this token once for the current session. After initial generation of this token, the value is stored in the session and is utilized for each subsequent request until the session expires. When a request is issued by the end-user, the server-side component must verify the existence and validity of the token in the request as compared to the token found in the session. If the token was not found within the request or the value provided does not match the value within the session, then the request should be aborted, token should be reset and the event logged as a potential CSRF attack in progress.